File: src/main/java/org/example/commons/repositories/api/UserRepository.java
package org.example.commons.repositories.api;

import org.example.commons.api.Filterable;
import org.example.commons.api.Repository;
import org.example.commons.entities.User;
import org.example.commons.entities.UserStatus;
import org.example.commons.entities.UserType;

import java.util.List;

public interface UserRepository extends Repository<User, Long>, Filterable<User> {
    List<User> findByType(UserType type);
    List<User> findByStatus(UserStatus status);
    List<User> findByTypeAndStatus(UserType type, UserStatus status);
}

File: src/main/kotlin/org/example/commons/repositories/api/UserRepository.kt
package org.example.commons.repositories.api

import org.example.commons.api.Filterable
import org.example.commons.api.Repository
import org.example.commons.entities.User
import org.example.commons.entities.UserRole
import org.example.commons.entities.UserStatus
import org.example.commons.entities.UserType

interface UserRepository : Repository<User, Long> {
    fun findByType(type: UserType): List<User>
    fun findByStatus(status: UserStatus): List<User>
    fun findByRole(role: UserRole): List<User>
    fun findByTypeAndStatus(type: UserType, status: UserStatus): List<User>
    fun findByTypeAndRole(type: UserType, role: UserRole): List<User>
    fun findByStatusAndRole(status: UserStatus, role: UserRole): List<User>
    fun findByTypeAndStatusAndRole(type: UserType, status: UserStatus, role: UserRole): List<User>
}